File: src/pkg_cpp_py/pkg_cpp_py/sub.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from interfaces.msg import Num

class Sub(Node) :
    def __init__(self) :
        super().__init__("subscriber")
        self.subscription_ = self.create_subscription(Num, "topic", self.callback_topic, 10)

    def callback_topic(self, msg) :
        self.get_logger().info("{}".format(msg.num))

def main(args=None) :
    rclpy.init(args=args)
    node = Sub()
    rclpy.spin(node)
    rclpy.shutdown()

if __name__ == "__main__" :
    main()

File: src/pkg_cpp_py/src/sub.cpp
#include "../include/pkg_cpp_py/sub.hpp"

Sub::Sub() : Node("subscriber")
{
    this->subscription_ = this->create_subscription<interfaces::msg::Num>("topic", 10, std::bind(&Sub::callback_topic, this, _1));
    RCLCPP_INFO(this->get_logger(), "Start subscriber");
}

void Sub::callback_topic(const interfaces::msg::Num &msg) const
{
    RCLCPP_INFO(this->get_logger(), "%ld", msg.num);
}

int main(int argc, char* argv[])
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<Sub>());
    rclcpp::shutdown();

    return 0;
}
